src/main/java/com/online/bookstore/service/impl/OrderServiceImpl.java

```diff
@@ -0,0 +1,126 @@
+    private final BookService bookService;
+    private final OrderMapper orderMapper;
+    private final OrderItemMapper orderItemMapper;
+
+    @Override
+    public List<OrderDto> getOrdersByUserId(final Long userId) {
+        return orderMapper.toDto(orderRepository.findAllByUserId(userId));
+    }
+
+    @Override
+    @Transactional
+    public OrderDto placeOrder(final Long userId,
+                               final CreateOrderRequestDto requestDto) {
+        final ShoppingCartDto cartByUserId = shoppingCartService.getCartByUserId(userId);
+        validateShoppingCart(cartByUserId);
+
+        final Set<OrderItem> orderItems = getOrderItems(cartByUserId.cartItems());
+        final Order order = getOrder(userId, requestDto.shippingAddress(), orderItems);
+        final Order savedOrder = orderRepository.save(order);
+
+        shoppingCartService.releaseCart(cartByUserId.id());
+
+        return orderMapper.toDto(savedOrder);
+    }
+
+    private static Order getOrder(final Long userId,
+                                  final String shippingAddress,
+                                  final Set<OrderItem> orderItems) {
+        final User user = new User().setId(userId);
+        return new Order()
+                .setUser(user)
+                .setStatus(OrderStatus.PENDING)
+                .setTotal(calculateOrderTotal(orderItems))
+                .setOrderDate(LocalDateTime.now(DEFAULT_ZONE_ID))
+                .setShippingAddress(shippingAddress)
+                .setOrderItems(orderItems);
+    }
+
+    private static void validateShoppingCart(ShoppingCartDto cartByUserId) {
+        if (cartByUserId.cartItems().isEmpty()) {
+            throw new SystemException("Cannot place an order, shopping cart is empty",
+                    HttpStatus.BAD_REQUEST);
+        }
+    }
+
+    private Set<OrderItem> getOrderItems(final List<CartItemDto> cartItems) {
+        final Map<Long, BigDecimal> bookPrices = getBookPrices(cartItems);
+        return cartItems.stream()
+                .map(cartItem -> orderItemMapper.toModel(
+                        cartItem,
+                        bookPrices.get(cartItem.bookId())
+                ))
+                .collect(Collectors.toSet());
+    }
+
+    private Map<Long, BigDecimal> getBookPrices(final List<CartItemDto> cartItems) {
+        final List<Long> bookIds = cartItems.stream()
+                .map(CartItemDto::bookId)
+                .toList();
+        return bookService.getBooksByIds(bookIds).stream()
+                .collect(Collectors.toMap(
+                        BookDto::id,
+                        BookDto::price
+                ));
+    }
+
+    private static BigDecimal calculateOrderTotal(final Set<OrderItem> orderItems) {
+        return orderItems.stream()
+                .map(orderItem -> orderItem.getPrice()
+                        .multiply(BigDecimal.valueOf(orderItem.getQuantity())))
+                .reduce(BigDecimal.ZERO, BigDecimal::add);
+    }
```

Comment:
I assure you it's not right to order a private method after its public method. Think of it like this: the next developer wants to get acquainted with this class. The first thing he thinks "What is the purpose of this class", so he looks at the state of the class (fields) and functionality (method). He doesn't care about the exact implementation of each method. In case, he wants to investigate some method, he'll look through it, and if he sees some private method he uses Ctrl + B, to the bottom of a class where each private method is implemented. https://stackoverflow.com/questions/1760850/best-practice-ordering-of-public-protected-private-within-the-class-definition
Also, look up some articles yourself

Reply:
reordered, thank you!